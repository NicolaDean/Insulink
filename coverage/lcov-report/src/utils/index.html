
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/utils</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">51.97% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>184/354</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">20.16% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>24/119</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>37/74</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">52.02% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>180/346</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="apiHelper.js"><a href="apiHelper.js.html">apiHelper.js</a></td>
	<td data-value="92.5" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 92%"></div><div class="cover-empty" style="width: 8%"></div></div>
	</td>
	<td data-value="92.5" class="pct high">92.5%</td>
	<td data-value="40" class="abs high">37/40</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="88.88" class="pct high">88.88%</td>
	<td data-value="9" class="abs high">8/9</td>
	<td data-value="92.5" class="pct high">92.5%</td>
	<td data-value="40" class="abs high">37/40</td>
	</tr>

<tr>
	<td class="file high" data-value="apiQuery.js"><a href="apiQuery.js.html">apiQuery.js</a></td>
	<td data-value="86.48" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 86%"></div><div class="cover-empty" style="width: 14%"></div></div>
	</td>
	<td data-value="86.48" class="pct high">86.48%</td>
	<td data-value="37" class="abs high">32/37</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="6" class="abs high">5/6</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="10" class="abs high">9/10</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="35" class="abs high">30/35</td>
	</tr>

<tr>
	<td class="file low" data-value="auth.js"><a href="auth.js.html">auth.js</a></td>
	<td data-value="5.55" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 5%"></div><div class="cover-empty" style="width: 95%"></div></div>
	</td>
	<td data-value="5.55" class="pct low">5.55%</td>
	<td data-value="36" class="abs low">2/36</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="14" class="abs low">0/14</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="5.55" class="pct low">5.55%</td>
	<td data-value="36" class="abs low">2/36</td>
	</tr>

<tr>
	<td class="file low" data-value="chartDataFormatter.js"><a href="chartDataFormatter.js.html">chartDataFormatter.js</a></td>
	<td data-value="5.26" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 5%"></div><div class="cover-empty" style="width: 95%"></div></div>
	</td>
	<td data-value="5.26" class="pct low">5.26%</td>
	<td data-value="19" class="abs low">1/19</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="5.26" class="pct low">5.26%</td>
	<td data-value="19" class="abs low">1/19</td>
	</tr>

<tr>
	<td class="file low" data-value="firebaseQuery.js"><a href="firebaseQuery.js.html">firebaseQuery.js</a></td>
	<td data-value="46.34" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 46%"></div><div class="cover-empty" style="width: 54%"></div></div>
	</td>
	<td data-value="46.34" class="pct low">46.34%</td>
	<td data-value="82" class="abs low">38/82</td>
	<td data-value="3.84" class="pct low">3.84%</td>
	<td data-value="26" class="abs low">1/26</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="15" class="abs medium">9/15</td>
	<td data-value="48.05" class="pct low">48.05%</td>
	<td data-value="77" class="abs low">37/77</td>
	</tr>

<tr>
	<td class="file low" data-value="inputChecker.js"><a href="inputChecker.js.html">inputChecker.js</a></td>
	<td data-value="20.37" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 20%"></div><div class="cover-empty" style="width: 80%"></div></div>
	</td>
	<td data-value="20.37" class="pct low">20.37%</td>
	<td data-value="54" class="abs low">11/54</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="36" class="abs low">0/36</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="20.37" class="pct low">20.37%</td>
	<td data-value="54" class="abs low">11/54</td>
	</tr>

<tr>
	<td class="file medium" data-value="insulineCalculator.js"><a href="insulineCalculator.js.html">insulineCalculator.js</a></td>
	<td data-value="57.14" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 57%"></div><div class="cover-empty" style="width: 43%"></div></div>
	</td>
	<td data-value="57.14" class="pct medium">57.14%</td>
	<td data-value="21" class="abs medium">12/21</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="12" class="abs medium">8/12</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="5" class="abs low">2/5</td>
	<td data-value="57.14" class="pct medium">57.14%</td>
	<td data-value="21" class="abs medium">12/21</td>
	</tr>

<tr>
	<td class="file medium" data-value="localStoreManager.js"><a href="localStoreManager.js.html">localStoreManager.js</a></td>
	<td data-value="77.77" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 77%"></div><div class="cover-empty" style="width: 23%"></div></div>
	</td>
	<td data-value="77.77" class="pct medium">77.77%</td>
	<td data-value="63" class="abs medium">49/63</td>
	<td data-value="57.14" class="pct medium">57.14%</td>
	<td data-value="14" class="abs medium">8/14</td>
	<td data-value="56.25" class="pct medium">56.25%</td>
	<td data-value="16" class="abs medium">9/16</td>
	<td data-value="77.41" class="pct medium">77.41%</td>
	<td data-value="62" class="abs medium">48/62</td>
	</tr>

<tr>
	<td class="file high" data-value="testingJsons.js"><a href="testingJsons.js.html">testingJsons.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at Thu May 26 2022 11:26:03 GMT+0200 (Central European Summer Time)
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    