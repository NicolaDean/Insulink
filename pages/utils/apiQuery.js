
import React from 'react';
import * as testingJson from "./testingJsons"

const debug = true;

export const imgUrl = "https://spoonacular.com/cdn/ingredients_100x100/";
const ingredientSearch = "";
const apiKey = "apiKey=916e4c93a5854139b950d4708c25bd80";
const apiRoot = "https://api.spoonacular.com/";
const apiIngr = "food/ingredients/";


/**
 * eg : https://api.spoonacular.com/food/ingredients/{id}/information
 * retrive details of ingredient with id
 * @param {*} id id of the ingredients
 * @returns details of ingredients
 */
export const getIngredientDetails = (id)=>{

    if(debug) return testingJson.foodDetails;

    let query = apiRoot + apiIngr + id +"/information?" +apiKey +"&amount=100&unit=grams";;
    console.log("QUERY: " + query);
    return getJson(query);
}

/**
 * eg : https://api.spoonacular.com/food/ingredients/search
 * get all ingredients with similar name of userInput
 * @param {*} userInput user query atribute
 * @returns a list of ingredients
 */
export const getIngredentList = (userInput) =>{

    if(debug){
        return testingJson.ingredientList;
    }

    let query = apiRoot + apiIngr + "search?query=" + userInput + "&number=4&" + apiKey
    console.log("QUERY: " + query);
    return getJson(query);
}

/**
 * given a nutrientsJson map each nutritions name to its contents/values
 * @param {*} nutrientsJson 
 * @returns a dictionary of nutrients
 */
export const extractNutrients = (nutrientsJson) =>{

    let nutrients = {};

    nutrientsJson.forEach( nut =>{
        nutrients[nut.name] = nut; //Map each nutrient name to its content
    });

    return nutrients;
}

export const extractProperties = (propertiesJson) =>{
    let properties = {};

    propertiesJson.forEach( p =>{
        properties[p.name] = p; //Map each properties name to its content
    })

    return properties;
}
/**
 * Do a fetch operation on the query given in input and return the json response from API
 * @param {*} query API QUERY
 * @returns 
 */
const getJson = async (query) =>{
    
    //Prepare request
    const response = fetch(query);
    //WAITING response from API
    const json = await(await response).json();

    if(json.code = 402){
        console.log(json)
        console.log("FINISHED DAILY API CALL");
    } 
    else console.log(json);
    //Return
    return json;
}