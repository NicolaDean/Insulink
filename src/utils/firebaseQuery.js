
import firestore from '@react-native-firebase/firestore'
import { firebase } from "@react-native-firebase/firestore";
//TABLES

const userTable = "Users";
const glicemyTable = "Glicemy";
const diaryTable = "FoodDiary";

const users = firestore().collection(userTable);

//USER QUERY:-------------------------------------------------------------------

class firebaseQuery{
        /**
     * get the user actual data using email as id
     * @param {*} email email to query
     * @returns user data
     */
    getUserData = async (email) =>{
        const user = (await users.doc(email).get()).data();

    // console.log("GLIC: " + JSON.stringify(glicemy))
        return user;
    }

    registerUser = async (id,userData) =>{
        await (users.doc(id).set(userData));
    }

    udpateUser = async (userData) =>{
        //TODO UPDATE USER DATA
    }
    //------------------------------------------------------------------------------------
    //GLICEMY QUERY:----------------------------------------------------------------------

    zeroPad = (num, places) => String(num).padStart(places, '0');

    /**
     * Get the glicemy data of a specific user
     * @param {*} email email that identify user
     * @returns an array containing the glicemy data
     */
    getUserGlicemy = async (userId,date=new Date()) =>{

        let glicemy_records = {};
        date = this.glicemyDateFormatter(date);
        glicemy_records[date] = [];
        //const today = "27-02-2022";
        const res = (await (users.doc(userId).collection(glicemyTable).doc(date).get())).data();
        
        if(res == undefined) return [];

        res.data.forEach(g => {
            glicemy_records[date].push(this.changeGlicemyTimeFormat(g));
        });

        console.log("->>>>" + JSON.stringify(glicemy_records));

        return glicemy_records;
    }

    /**
     * Add a new value of glicemy to user with userId
     * @param {*} userId identifier of user
     * @param {*} glicemyData data to add
     */
    addGlicemyValue = async (userId,glicemyData) =>{

        const id = this.glicemyDateFormatter();
        const new_glicemy = await users.doc(userId).collection(glicemyTable).doc(id).get();
        
        if(new_glicemy && new_glicemy.exists)
        {
            await new_glicemy.ref.update({
                data:firebase.firestore.FieldValue.arrayUnion(glicemyData)
            })
        }else{
            await new_glicemy.ref.set({data:[glicemyData]});
        }
        

    }

    glicemyDateFormatter = (date = new Date()) =>
    {
        const today = date;
        let id = this.zeroPad(today.getDate(),2) +"-"+ this.zeroPad(today.getMonth()+1,2) +"-"+ today.getFullYear();

        return id;
    }

    getTodayGlicemy = (glicemy) =>
    {
        const today = this.glicemyDateFormatter();

        if(glicemy==null) return null;
        
        return glicemy[today];
    }

    getTodayLastGlicemy =(glicemy) =>
    {
        if(glicemy==null) {return null};
        const today = this.glicemyDateFormatter();

        const todayGlycemia= glicemy[today];
        
        if(todayGlycemia!=undefined){
            const len=Object.keys(todayGlycemia).length;
            return todayGlycemia[len-1].value;
        }
        else return 120;
    }


    changeGlicemyTimeFormat = (glicemy,realdate=false) => 
    {
        let time = {};
        if(realdate)
        {
            time = glicemy.time;
        }
        else{
            time = new Date(glicemy.time.seconds*1000);
        }
        

        let date = this.zeroPad(time.getDate(),2) +"/"+ this.zeroPad(time.getMonth(),2);
        let hours = this.zeroPad(time.getHours(),2) +":"+ this.zeroPad(time.getMinutes(),2);

        //TODO ERROR OF DATE FORMAT CHECK!
        glicemy.time = {date:date,hours:hours};
        return glicemy;
    }

    saveFoodDiary = async (userId,date,diary) =>{
        console.log("TRY FIREBASE")
        console.log("USER : " + userId + "->" + date);
        const userDiary = await users.doc(userId).collection(diaryTable).doc(date).get();
        firestore().settings({ ignoreUndefinedProperties: true }); //INGORE UNDEFINED FIELD
        const data = {
            totMacro:diary.totMacro,
            meals:diary.meals,
            activities:diary.activities
        }

        //await users.doc(userId).collection(diaryTable).doc(date).set({data:diary},{merge: true});
       if(userDiary && userDiary.exists)
        {
            console.log("exist");
            try{
                await userDiary.ref.update(data);
            }catch(e){
                console.log("ERRORE : " + e);
            }
            
        }else{
            
            await userDiary.ref.set(data)//TODO VEDERE CHE SUCCEDE QUI 
            console.log("Saved on " + userId + "-> " + date);
        }
    }

    getFoodDiary = async (userId,date) =>{
        const res = (await (users.doc(userId).collection(diaryTable).doc(date).get())).data();
        
        if(res == undefined) return [];

        return res;
    }
}

export const FirebaseQuery = new firebaseQuery(); 