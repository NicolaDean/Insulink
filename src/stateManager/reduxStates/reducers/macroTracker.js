
import { macroConstants } from "../../../constants/states";
import { foodMethods } from "../../../constants/reducers"
import { FirebaseQuery } from "../../../utils/firebaseQuery";

const macro = macroConstants;

//TODO Each diary will have a Date, if Date < Today it will be saved localy/firabase and resetted for new day
export const initialDiaryState = {
    currentMeal:"breakfast",
    currentDate:FirebaseQuery.glicemyDateFormatter(new Date()),
    totMacro:{cal:0,carb:0,fat:0,prot:0},
    id:1,
    activity_id:1,
    meals:{
        breakfast:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}},
        lunch:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}},
        dinner:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}},
        snack:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}}
    },
    history:{
        empty:true,
        totMacro:{cal:0,carb:0,fat:0,prot:0},
        meals:{
            breakfast:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}},
            lunch:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}},
            dinner:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}},
            snack:{foods:[],macro:{cal:0,carb:0,fat:0,prot:0}}
        },
        activities:{
            breakfast:{sports:[],totCal:0},
            lunch:{sports:[],totCal:0},
            dinner:{sports:[],totCal:0},
            snack:{sports:[],totCal:0}
        }
    },
    totCalBurned:0,
    activities:{
        breakfast:{sports:[],totCal:0},
        lunch:{sports:[],totCal:0},
        dinner:{sports:[],totCal:0},
        snack:{sports:[],totCal:0}
    }
}

/**
 * Select current meal to assign added food to
 * @param {*} state 
 * @param {*} data 
 * @returns 
 */
const selectMeal = (state,data)=>{
    const newState = {
        ...state,
        currentMeal: data.mealType}

    console.log("Current Meal:" + state.currentMeal);
    return newState;
}

/**
 * Merge 2 macro into a single one by summing its values
 * @param {*} base first macro
 * @param {*} toSum second macro
 * @returns merged macros
 */
const sumMacro = (base,toSum) =>{

    let result = {};

    result[macro.cal] = base[macro.cal] + toSum[macro.cal];
    result[macro.fat] = base[macro.fat] + toSum[macro.fat];
    result[macro.carb] = base[macro.carb] + toSum[macro.carb];
    result[macro.prot] = base[macro.prot] + toSum[macro.prot];

    console.log("ADDING THIS MACRO: " + base[macro.cal] +  " + " + toSum[macro.cal] +"=" +  result[macro.cal]);
   
    return result;
}
/**
 * Merge 2 macro into a single one by subbing its values
 * @param {*} base first macro
 * @param {*} toSum second macro
 * @returns merged macros
 */
const subMacro = (base,toRemove) =>{

    let result = {};

    
    result[macro.cal] = base[macro.cal] - toRemove[macro.cal];
    result[macro.fat] = base[macro.fat] - toRemove[macro.fat];
    result[macro.carb] = base[macro.carb] - toRemove[macro.carb];
    result[macro.prot] = base[macro.prot] - toRemove[macro.prot];

    console.log("REMOVING THIS MACRO: " + base[macro.cal] +  " - " + toRemove[macro.cal] +"=" +  result[macro.cal]);
    return result;
}

/**
 * add a food to the global state representing meals of today
 * @param {*} state today meals
 * @param {*} data food to add
 * @returns new state with the new food added to the current selected meal
 */
const addFood = (state,data)=>{
    
    const food = data.food;
    //COPY STATE
    const newstate ={...state};
    food.identifier = newstate.id;
    //INCREMENT ID
    newstate.id = newstate.id + 1;
    //PUSH FOOD

    newstate.meals[state.currentMeal].foods.push(data.food);
    //CALCULATE TOT MACRO
    let m = state.meals[state.currentMeal].macro;
    newstate.totMacro  = sumMacro(newstate.totMacro,data.food);
    newstate.meals[state.currentMeal].macro = sumMacro(m,data.food);

    return newstate;
}

const addActivity = (state,data) =>{
    const activity = data.activity;

    //COPY STATE
    const newstate ={...state};
    activity.identifier = newstate.activity_id;
    newstate.activity_id = newstate.activity_id+1;
    //
    newstate.activities[newstate.currentMeal].sports.push(activity);
    newstate.activities[newstate.currentMeal].totCal += activity.calories;
    newstate.totCalBurned += activity.calories;

    return newstate;
}



const findMatchingIdMeal = (array,identifier,isSport = false) =>{

    console.log("Identifier: " + identifier)
    //REMOVING FOOD FROM LIST:
    let found = false;
    let m_found = false;

    let meal_types = Object.keys(array);

    //FOR EACH MEAL
    meal_types.forEach((key,index) =>{
        if(found != false){ return;}
            let i = 0;
            let tmp = [];
            if(isSport) {
                tmp =array[key].sports;
            }else{
                tmp =array[key].foods;
            }
            
    
            //FOR EACH FOOD
            tmp.forEach(x =>{
                //CHECK IDENTIFIER TO REMOVE
                console.log("(" + x.identifier + " , " + identifier+ ")");
            
                if(x.identifier === identifier)
                {
                    //IF FOUNDED MARK INDEX AND MEAL FROM WICH REMOVE
                    found = i;
                    m_found = key;
                }
                i++;
            });        
        });

        return {index:found,meal:m_found};
}
/**
 * remove a food to the global state representing meals of today
 * @param {*} state today meals
 * @param {*} data food to add
 * @returns new state with the selected food removed to the current selected meal
 */
const removeFood = (state,data)=>{

    const food = data.food;
    //COPY STATE
    const newstate ={...state};
   
    console.log("FOOOD: " + JSON.stringify(food));

    //FIND MATCHING FOOD IN STATE
    let matching = findMatchingIdMeal(newstate.meals,food.identifier,false);

    //REMOVE FROM MEAL THE FOOD AT THE SPECIFIC INDEX MARKED
    if((matching.index-1 === -1) || matching.index != false){
        console.log("REMOVING: " + JSON.stringify(food));

        //REMOVING MACRO
        let m = state.meals[matching.meal].macro;
        newstate.meals[matching.meal].macro = subMacro(m,food);
        newstate.totMacro  = subMacro(newstate.totMacro,food);

        //REMOVING FOOD FROM LIST
        newstate.meals[matching.meal].foods.splice(matching.index,1);
        
    }
    

    return newstate;
}

const editFood = (state,data) =>{
    const food = data.food;
    //COPY STATE
    const newstate ={...state};
   
    //FIND MATCHING FOOD IN STATE
    let matching = findMatchingIdMeal(newstate.meals,food.identifier,false);


    //REMOVE FROM MEAL THE FOOD AT THE SPECIFIC INDEX MARKED
    if((matching.index-1 === -1) || matching.index != false){
        console.log("FOUND IN (" +matching.meal + ","+matching.index+")" );
        //PROBLEM IS IN IDENTIFIER!!
        let m = newstate.meals[matching.meal].macro;
        console.log(JSON.stringify(newstate));
        let oldFood = newstate.meals[matching.meal].foods[matching.index];

        //REMOVE FOOD FROM MACRO
        //(BISOGNA PASSARE IL CIBO VERO E PROPRIO MO VIENE PASSATO L'ID)
        //Remove old version Macro
        newstate.meals[matching.meal].macro = subMacro(m,oldFood);
        newstate.totMacro  = subMacro(newstate.totMacro,oldFood);

        //Add new Version Macro
        newstate.meals[matching.meal].macro = sumMacro(newstate.meals[matching.meal].macro,food);
        newstate.totMacro  = sumMacro(newstate.totMacro,food);

        newstate.meals[matching.meal].foods[matching.index] = food;
    }
    

    return newstate;
}

const removeActivity = (state,data) =>{
    //COPY STATE
    const newstate ={...state};
    const activity = data.activity;


    //FIND MATCHING ACTIVITY IN STATE
    let matching = findMatchingIdMeal(newstate.activities,activity.identifier,true);


    if((matching.index-1 === -1) || matching.index != false){
        //TODO REMOVE CALORIES
        newstate.activities[matching.meal].totCal -= activity.calories;
        newstate.totCalBurned -= activity.calories;
        //REMOVING FOOD FROM LIST
        newstate.activities[matching.meal].sports.splice(matching.index,1);
    }
    //TODO REMOVE ACTIVITY
    return newstate;
}
/**
 * Load Meals data from memory
 * @param {*} state 
 * @param {*} payload 
 * @returns 
 */
const loadMeals = (state,payload) =>{
    return payload.diary;
}

/**
 * reset the diary for a new day
 * @param {*} state 
 * @param {*} payload 
 * @returns 
 */
const resetDiary = (state,payload) =>{

    const newState = {...initialDiaryState};

    console.log("UFF: " + JSON.stringify(newState))

    newState.currentDate = payload.currentDate;

    return newState;
}

const loadHistory = (state,payload) => {
    const newState = {...state};

    newState.currentDate = payload.date;
    if(payload.date != FirebaseQuery.glicemyDateFormatter()){
        if(payload.history!=null){
            newState.history = payload.history;
        }
    }
    
    return newState;
}

//TODO PUT THE NAME OF ACTIONS INTO COSNTANT FILE
const macroReducer = (state = initialDiaryState, action) => {
    switch(action.type){
        case foodMethods.addFood:
            return addFood(state,action.payload);
        case foodMethods.removeFood:
            return removeFood(state,action.payload);
        case foodMethods.editFood:
            return editFood(state,action.payload);
        case foodMethods.selectMeal:
            return selectMeal(state,action.payload);
        case foodMethods.loadFoodDiary:
            return loadMeals(state,action.payload);
        case foodMethods.addActivity:
            return addActivity(state,action.payload);
        case foodMethods.resetDiary:
            return resetDiary(state,action.payload);
        case foodMethods.removeActivity:
            return removeActivity(state,action.payload);
        case foodMethods.loadHistory:
            return loadHistory(state,action.payload);
        default:
            return state;
        }
}

export default macroReducer;