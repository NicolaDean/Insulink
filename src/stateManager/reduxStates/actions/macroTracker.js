import { foodMethods } from "../../../constants/reducers"

export const addFood = (food) =>{
    return{
        type: foodMethods.addFood,
        payload: {food:food}
    }
} 

export const removeFood = (food,mealType) =>{
    return{
        type: foodMethods.removeFood,
        payload: {food:food,mealType:mealType}
    }
} 

export const selectMealType = (mealType) =>{
    return{
        type: foodMethods.selectMeal,
        payload: {mealType:mealType}
    }
}