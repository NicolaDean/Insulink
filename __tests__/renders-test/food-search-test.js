import React from "react";
//import {cleanup, fireEvent, render} from '@testing-library/react';
import {create,act} from 'react-test-renderer';
import CustomImageButton from "../../src/customComponents/customImageButton";
import FoodSearch from "../../src/pages/foodSearch/foodSearch";
import * as actions from "../../src/stateManager/reduxStates/actions/macroTracker";
import { Food_API } from "../../src/utils/apiQuery";
import { FirebaseQuery } from "../../src/utils/firebaseQuery";
import { dummyApple, dummyAppleDeletable } from "../../testHelper/dataForTest/foods";
import CustomFirestoreMock from "../../__mocks__/@react-native-firebase/firebase";
jest.useFakeTimers();
const mock_firebase = new CustomFirestoreMock();

const firebasee = () =>{
    return {
        collection:jest.fn((a)=>{console.log("COLLECTION->"+ a);return mock_firebase}),
        doc:jest.fn((a)=>{console.log("DOC->"+ a);return mock_firebase})
    };
}
jest.mock('../../src/utils/firebaseQuery');
jest.mock('@react-native-firebase/firestore',()=> firebasee);
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.mock('react-native/Libraries/Animated/Easing');
jest.mock('react-native/Libraries/Animated/animations/TimingAnimation');



const mock_navigation = {
    navigate:jest.fn((a,b)=>{console.log("NAV to " + a)})
}

const mockApi = jest.spyOn(Food_API,'getFoodListAlternative');
mockApi.mockImplementation = (a) =>{console.log("METHOD" +  a)}
const tree = create(<FoodSearch nav={mock_navigation}/>);

describe("TEST ON FOOD COMPONENT:",()=>{

    /**
     * Semplicemente snapshot
     */
    test("Food SnapShot: ",()=>{
        expect(tree).toMatchSnapshot();
    });

    /**
     * Test if Search button trigger API call
     */
    test("Search Button Click", () =>{
        //Search the Image search button
        const barrcodeBtn = tree.root.findAllByType(CustomImageButton)[0].props;
        //Click it
        act(()=>barrcodeBtn.onPress());
        //Check if navigation is called correctly
    });

    test("Barrcode Button Click", () =>{
        //Search the Image search button
        const barrcodeBtn = tree.root.findAllByType(CustomImageButton)[1].props;
        //Click it
        act(()=>barrcodeBtn.onPress());
        //Check if navigation is called correctly
        expect(mock_navigation.navigate).toBeCalledWith('ScannerPage',{});
    })

    
});