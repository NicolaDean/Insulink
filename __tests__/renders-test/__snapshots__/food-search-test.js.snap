// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST ON FOOD COMPONENT: Food SnapShot:  1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-around",
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.23,
        "shadowRadius": 2.62,
      }
    }
  >
    <TextInput
      adjustsFontSizeToFit={true}
      onChangeText={[Function]}
      placeholder="insert food!"
      style={
        Object {
          "fontSize": 31.25,
          "marginLeft": 15,
          "width": 375,
        }
      }
    />
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="ImageButton"
    >
      <View
        style={null}
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/search.png",
            }
          }
          style={
            Object {
              "height": 160.07999999999998,
              "marginLeft": 20,
              "resizeMode": "contain",
              "width": 90,
            }
          }
        />
      </View>
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="ImageButton"
    >
      <View
        style={null}
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/camera.png",
            }
          }
          style={
            Object {
              "height": 160.07999999999998,
              "marginLeft": 20,
              "resizeMode": "contain",
              "width": 90,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignContent": "center",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "alignItems": "center",
          "paddingVertical": 0,
        }
      }
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
